File: 2762-Continuous Subarrays.py
# # sortedList
from sortedcontainers import SortedList
class Solution:
    def continuousSubarrays(self, nums: List[int]) -> int:
        res = 0
        left = 0
        N = len(nums)
        sl = SortedList()
        
        for i in range(N):
            sl.add(nums[i])
            while sl[-1] - sl[0] > 2:
                sl.remove(nums[left])
                left += 1
            res += (i-left+1)
        
        return res

# deque
class Solution:
    def continuousSubarrays(self, nums: List[int]) -> int:
        maxQ, minQ = deque(), deque()
        N = len(nums)
        left = res = 0
        
        for i in range(N):
            
            # keep all indexes and numbers greater or equal to nums[i], monotonically decreasing
            while maxQ and nums[maxQ[-1]] < nums[i]:
                maxQ.pop()
            maxQ.append(i)
            
            # keep all indexes and numbers smaller or equal to nums[i], monotonically increasing
            while minQ and nums[minQ[-1]] > nums[i]:
                minQ.pop()
            minQ.append(i)            
            
            while nums[maxQ[0]] - nums[minQ[0]] > 2:
                if maxQ[0] < minQ[0]:
                    left = maxQ[0] + 1
                    maxQ.popleft()
                else:
                    left = minQ[0] + 1
                    minQ.popleft()
            
            res += (i-left+1)
            
        return res

# heap
class Solution:
    def continuousSubarrays(self, nums: List[int]) -> int:
        maxHeap, minHeap = [], []
        N = len(nums)
        left = res = 0
        
        for i in range(N):
            heappush(maxHeap, (-nums[i], i))
            heappush(minHeap, (nums[i], i))
            while -maxHeap[0][0] - minHeap[0][0] > 2:
                left += 1
                while maxHeap and maxHeap[0][1] < left:
                    heappop(maxHeap)
                while minHeap and minHeap[0][1] < left:
                    heappop(minHeap)
            
            res += (i - left + 1)
            
        return res

File: C++/2762-Continuous Subarrays.cpp
class Solution {
public:
    long long continuousSubarrays(vector<int>& nums) {
        int N = nums.size();
        multiset<int> s;
        int left = 0;
        long long res = 0;

        for (int i=0; i<N; i++) {
            s.insert(nums[i]);
            while (*s.rbegin() - *s.begin() > 2) {
                s.erase(s.find(nums[left++]));
            }
            res += s.size();
        }

        return res;
    }
};



using PII = pair<int, int>;
class Solution {
public:
    long long continuousSubarrays(vector<int>& nums) {
        int N = nums.size();
        priority_queue<PII, vector<PII>, greater<>> minHeap;
        priority_queue<PII> maxHeap;
        int left = 0;
        long long res=0;
        for (int i=0; i<N; i++) {
            minHeap.push({nums[i], i});
            maxHeap.push({nums[i], i});
            while (maxHeap.top().first - minHeap.top().first > 2) {
                left++;
                while (!maxHeap.empty() && maxHeap.top().second < left) {
                    maxHeap.pop();
                }
                while (!minHeap.empty() && minHeap.top().second < left) {
                    minHeap.pop();
                }
            }
            res += (i-left+1);
        }
        return res;
    }
};


// Use deque to create monotonically increasing/decreasing deque
using LL = long long;
class Solution {
public:
    long long continuousSubarrays(vector<int>& nums) {
        LL res = 0;
        int N = nums.size();
        deque<int> maxD, minD;
        int left = 0;
        for (int i=0; i<N; i++) {
            while (!maxD.empty() && nums[maxD.back()] <= nums[i]) maxD.pop_back();  // monotonically increasing
            while (!minD.empty() && nums[minD.back()] >= nums[i]) minD.pop_back();  // monotonically decreasing
            maxD.push_back(i);
            minD.push_back(i);

            while (nums[maxD.front()] - nums[minD.front()] > 2) {
                left++;
                while (maxD.front() < left) maxD.pop_front();
                while (minD.front() < left) minD.pop_front();
            }
            res += i-left+1;
        }

        return res;
    }
};